After relocation shards might temporarily not be searchable if still in POST_RECOVERY

SimpleSortTests.testIssue8226 for example fails about once a week. Example failure:
http://build-us-00.elasticsearch.org/job/es_g1gc_1x_metal/3129/

I can reproduce it locally (although very rarely) with some additional logging (action.search.type: TRACE). 

Here is a brief analysis of what happened. Would be great if someone could take a look and let me know if this makes sense.

Failure:

```
1> REPRODUCE WITH  : mvn clean test -Dtests.seed=774A2866F1B6042D -Dtests.class=org.elasticsearch.search.sort.SimpleSortTests -Dtests.method="testIssue8226 {#76 seed=[774A2866F1B6042D:ACB4FF9F8C8CA341]}" -Des.logger.level=DEBUG -Des.node.mode=network -Dtests.security.manager=true -Dtests.nightly=false -Dtests.client.ratio=0.0 -Dtests.heap.size=512m -Dtests.jvm.argline="-server -XX:+UseConcMarkSweepGC -XX:-UseCompressedOops -XX:+AggressiveOpts -Djava.net.preferIPv4Stack=true" -Dtests.locale=fi_FI -Dtests.timezone=Etc/GMT+9 -Dtests.processors=4
  1> Throwable:
  1> java.lang.AssertionError: One or more shards were not successful but didn't trigger a failure
  1> Expected: <47>
  1>      but: was <46>
```

Here is an example failure in detail, the relevant parts of the logs are below:

State
-----

node_0 is master.
[test_5][0] is relocating from node_1 to node_0.
Cluster state 3673 has the shard as relocating, in cluster state 3674 it is started.
node_0 is the coordinating node for the search request.

In brief, the request fails for shard [test_5][0] because node_0 operates on an older cluster state 3673 when processing the search request, while node_1 is already on 3674.


Course of events:
----

1. node_0 sends shard started, but the shard is still in state POST_RECOVERY and will remain so until it receives the new cluster state and applies it locally
2. node_0(master) receives the shard started request and publishes the new cluster state 3674 to node_0 and node_1
3. node_1 receives the cluster state 3674 and applies it locally
4. node_0 sends search request for [test_5][0] to node_1 because according to cluster state 3673 the shard is there and relocating
 -> request fails with IndexShardMissingException because node_1 already applied cluster state 3674 and deleted the shard.
5. node_0 then sends request for [test_5][0] to node_0 because the shard is there as well (according to cluster state 3673 it is and initializing)
 -> request fails with IllegalIndexShardStateException because node_0 has not yet processed cluster state 3674 and therefore the shard is in POST_RECOVERY instead of STARTED
No shard failure is logged because IndexShardMissingException and IllegalIndexShardStateException are explicitly excluded from shard failures.
6. node_0 finally also gets to process the new cluster state and moves the shard [test_5][0] to STARTED but it is too late


This is a very rare condition and maybe too bad on client side because the information that one shard did not deliver results is there although it is not explicitly listed as shard failure. We can probably make the test pass easily be just waiting for relocations before executing the search request but that seems wrong because any search request can fail this way.


Sample log
----

```
[....]

  1> [2015-01-26 09:27:14,435][DEBUG][indices.recovery         ] [node_0] [test_5][0] recovery completed from [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}], took [84ms]
  1> [2015-01-26 09:27:14,435][DEBUG][cluster.action.shard     ] [node_0] sending shard started for [test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING], indexUUID [E3T8J7CaRkyK533W0hMBPw], reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]
  1> [2015-01-26 09:27:14,435][DEBUG][cluster.action.shard     ] [node_0] received shard started for [test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING], indexUUID [E3T8J7CaRkyK533W0hMBPw], reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]
  1> [2015-01-26 09:27:14,436][DEBUG][cluster.service          ] [node_0] processing [shard-started ([test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]), reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]]: execute
  1> [2015-01-26 09:27:14,436][DEBUG][cluster.action.shard     ] [node_0] [test_5][0] will apply shard started [test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING], indexUUID [E3T8J7CaRkyK533W0hMBPw], reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]


[....]


  1> [2015-01-26 09:27:14,441][DEBUG][cluster.service          ] [node_0] cluster state updated, version [3674], source [shard-started ([test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]), reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]]
  1> [2015-01-26 09:27:14,441][DEBUG][cluster.service          ] [node_0] publishing cluster state version 3674
  1> [2015-01-26 09:27:14,442][DEBUG][discovery.zen.publish    ] [node_1] received cluster state version 3674
  1> [2015-01-26 09:27:14,443][DEBUG][cluster.service          ] [node_1] processing [zen-disco-receive(from master [[node_0][GQ6yYxmyRT-sfvT0cmuqQQ][schmusi][inet[/192.168.2.102:9400]]{mode=network, enable_custom_paths=true}])]: execute
  1> [2015-01-26 09:27:14,443][DEBUG][cluster.service          ] [node_1] cluster state updated, version [3674], source [zen-disco-receive(from master [[node_0][GQ6yYxmyRT-sfvT0cmuqQQ][schmusi][inet[/192.168.2.102:9400]]{mode=network, enable_custom_paths=true}])]
  1> [2015-01-26 09:27:14,443][DEBUG][cluster.service          ] [node_1] set local cluster state to version 3674
  1> [2015-01-26 09:27:14,443][DEBUG][indices.cluster          ] [node_1] [test_5][0] removing shard (not allocated)
  1> [2015-01-26 09:27:14,443][DEBUG][index                    ] [node_1] [test_5] [0] closing... (reason: [removing shard (not allocated)])
  1> [2015-01-26 09:27:14,443][INFO ][test.store               ] [node_1] [test_5][0] Shard state before potentially flushing is STARTED
  1> [2015-01-26 09:27:14,453][DEBUG][search.sort              ] cluster state:
  1> version: 3673
  1> meta data version: 2043
  1> nodes:
  1>    [node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}
  1>    [node_0][GQ6yYxmyRT-sfvT0cmuqQQ][schmusi][inet[/192.168.2.102:9400]]{mode=network, enable_custom_paths=true}, local, master
  1> routing_table (version 3006):
  1> -- index [test_4]
  1> ----shard_id [test_4][4]
  1> --------[test_4][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][7]
  1> --------[test_4][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][7], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][0]
  1> --------[test_4][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][3]
  1> --------[test_4][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][1]
  1> --------[test_4][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][5]
  1> --------[test_4][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][6]
  1> --------[test_4][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][6], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][2]
  1> --------[test_4][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_3]
  1> ----shard_id [test_3][2]
  1> --------[test_3][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_3][0]
  1> --------[test_3][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][3]
  1> --------[test_3][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][1]
  1> --------[test_3][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][5]
  1> --------[test_3][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][6]
  1> --------[test_3][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_3][4]
  1> --------[test_3][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_2]
  1> ----shard_id [test_2][2]
  1> --------[test_2][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_2][0]
  1> --------[test_2][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_2][3]
  1> --------[test_2][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][1]
  1> --------[test_2][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][1], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][5]
  1> --------[test_2][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][4]
  1> --------[test_2][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_1]
  1> ----shard_id [test_1][0]
  1> --------[test_1][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_1][1]
  1> --------[test_1][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_0]
  1> ----shard_id [test_0][4]
  1> --------[test_0][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][0]
  1> --------[test_0][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][0], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][7]
  1> --------[test_0][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][7], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][3]
  1> --------[test_0][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][1]
  1> --------[test_0][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][5]
  1> --------[test_0][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][6]
  1> --------[test_0][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][2]
  1> --------[test_0][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_8]
  1> ----shard_id [test_8][0]
  1> --------[test_8][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_8][1]
  1> --------[test_8][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_8][2]
  1> --------[test_8][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1>
  1> -- index [test_7]
  1> ----shard_id [test_7][2]
  1> --------[test_7][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_7][0]
  1> --------[test_7][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_7][3]
  1> --------[test_7][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_7][1]
  1> --------[test_7][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_7][4]
  1> --------[test_7][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_6]
  1> ----shard_id [test_6][0]
  1> --------[test_6][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][3]
  1> --------[test_6][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][1]
  1> --------[test_6][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][2]
  1> --------[test_6][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_6][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_5]
  1> ----shard_id [test_5][0]
  1> --------[test_5][0], node[G4AEDzbrRae5BC_UD9zItA], relocating [GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[RELOCATING]
  1> ----shard_id [test_5][3]
  1> --------[test_5][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_5][1]
  1> --------[test_5][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_5][2]
  1> --------[test_5][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1>
  1> routing_nodes:
  1> -----node_id[GQ6yYxmyRT-sfvT0cmuqQQ][V]
  1> --------[test_4][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_1][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_8][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_8][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_7][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_7][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_6][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]
  1> --------[test_5][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> -----node_id[G4AEDzbrRae5BC_UD9zItA][V]
  1> --------[test_4][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][7], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][6], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_3][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_3][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_2][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_2][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][1], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_1][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][0], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][7], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_8][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_5][0], node[G4AEDzbrRae5BC_UD9zItA], relocating [GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[RELOCATING]
  1> --------[test_5][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_5][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ---- unassigned
  1>
  1> tasks: (1):
  1> 13638/URGENT/shard-started ([test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]), reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]/13ms
  1>



[...]


 1> [2015-01-26 09:27:14,459][TRACE][action.search.type       ] [node_0] got first-phase result from [G4AEDzbrRae5BC_UD9zItA][test_3][3]
  1> [2015-01-26 09:27:14,460][TRACE][action.search.type       ] [node_0] [test_5][0], node[G4AEDzbrRae5BC_UD9zItA], relocating [GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[RELOCATING]: Failed to execute [org.elasticsearch.action.search.SearchRequest@1469040a] lastShard [false]
  1> org.elasticsearch.transport.RemoteTransportException: [node_1][inet[/192.168.2.102:9401]][indices:data/read/search[phase/dfs]]
  1> Caused by: org.elasticsearch.index.IndexShardMissingException: [test_5][0] missing
  1>     at org.elasticsearch.index.IndexService.shardSafe(IndexService.java:203)
  1>     at org.elasticsearch.search.SearchService.createContext(SearchService.java:539)
  1>     at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:523)
  1>     at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:208)
  1>     at org.elasticsearch.search.action.SearchServiceTransportAction$SearchDfsTransportHandler.messageReceived(SearchServiceTransportAction.java:757)
  1>     at org.elasticsearch.search.action.SearchServiceTransportAction$SearchDfsTransportHandler.messageReceived(SearchServiceTransportAction.java:748)
  1>     at org.elasticsearch.transport.netty.MessageChannelHandler$RequestHandler.doRun(MessageChannelHandler.java:275)
  1>     at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:36)
  1>     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  1>     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  1>     at java.lang.Thread.run(Thread.java:745)
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_6][1]
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_5][2]
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_7][1]
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_3][2]
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_8][2]
  1> [2015-01-26 09:27:14,455][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_1][0]


[...]


  1> [2015-01-26 09:27:14,463][TRACE][action.search.type       ] [node_0] [test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]: Failed to execute [org.elasticsearch.action.search.SearchRequest@1469040a] lastShard [true]
  1> org.elasticsearch.index.shard.IllegalIndexShardStateException: [test_5][0] CurrentState[POST_RECOVERY] operations only allowed when started/relocated
  1>     at org.elasticsearch.index.shard.IndexShard.readAllowed(IndexShard.java:839)
  1>     at org.elasticsearch.index.shard.IndexShard.acquireSearcher(IndexShard.java:651)
  1>     at org.elasticsearch.index.shard.IndexShard.acquireSearcher(IndexShard.java:647)
  1>     at org.elasticsearch.search.SearchService.createContext(SearchService.java:543)
  1>     at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:523)
  1>     at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:208)
  1>     at org.elasticsearch.search.action.SearchServiceTransportAction$3.call(SearchServiceTransportAction.java:197)
  1>     at org.elasticsearch.search.action.SearchServiceTransportAction$3.call(SearchServiceTransportAction.java:194)
  1>     at org.elasticsearch.search.action.SearchServiceTransportAction$23.run(SearchServiceTransportAction.java:559)
  1>     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  1>     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  1>     at java.lang.Thread.run(Thread.java:745)
  1> [2015-01-26 09:27:14,459][TRACE][action.search.type       ] [node_0] got first-phase result from [G4AEDzbrRae5BC_UD9zItA][test_2][4]



[...]



  1> [2015-01-26 09:27:14,493][DEBUG][cluster.service          ] [node_0] set local cluster state to version 3674
  1> [2015-01-26 09:27:14,493][DEBUG][index.shard              ] [node_0] [test_5][0] state: [POST_RECOVERY]->[STARTED], reason [global state is [STARTED]]
  1> [2015-01-26 09:27:14,493][DEBUG][river.cluster            ] [node_0] processing [reroute_rivers_node_changed]: execute
  1> [2015-01-26 09:27:14,493][DEBUG][river.cluster            ] [node_0] processing [reroute_rivers_node_changed]: no change in cluster_state
  1> [2015-01-26 09:27:14,493][DEBUG][cluster.service          ] [node_0] processing [shard-started ([test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], relocating [G4AEDzbrRae5BC_UD9zItA], [P], s[INITIALIZING]), reason [after recovery (replica) from node [[node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}]]]: done applying updated cluster_state (version: 3674)
  1> [2015-01-26 09:27:14,456][TRACE][action.search.type       ] [node_0] got first-phase result from [GQ6yYxmyRT-sfvT0cmuqQQ][test_2][3]


[...]

  1> [2015-01-26 09:27:14,527][DEBUG][search.sort              ] cluster state:
  1> version: 3674
  1> meta data version: 2043
  1> nodes:
  1>    [node_1][G4AEDzbrRae5BC_UD9zItA][schmusi][inet[/192.168.2.102:9401]]{mode=network, enable_custom_paths=true}
  1>    [node_0][GQ6yYxmyRT-sfvT0cmuqQQ][schmusi][inet[/192.168.2.102:9400]]{mode=network, enable_custom_paths=true}, local, master
  1> routing_table (version 3007):
  1> -- index [test_4]
  1> ----shard_id [test_4][2]
  1> --------[test_4][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][7]
  1> --------[test_4][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][7], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][0]
  1> --------[test_4][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][3]
  1> --------[test_4][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][1]
  1> --------[test_4][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][5]
  1> --------[test_4][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_4][6]
  1> --------[test_4][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][6], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_4][4]
  1> --------[test_4][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_3]
  1> ----shard_id [test_3][4]
  1> --------[test_3][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_3][0]
  1> --------[test_3][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][3]
  1> --------[test_3][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][1]
  1> --------[test_3][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][5]
  1> --------[test_3][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_3][6]
  1> --------[test_3][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_3][2]
  1> --------[test_3][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_2]
  1> ----shard_id [test_2][4]
  1> --------[test_2][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_2][0]
  1> --------[test_2][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_2][3]
  1> --------[test_2][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][1]
  1> --------[test_2][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][1], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][5]
  1> --------[test_2][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_2][2]
  1> --------[test_2][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_1]
  1> ----shard_id [test_1][0]
  1> --------[test_1][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_1][1]
  1> --------[test_1][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_0]
  1> ----shard_id [test_0][2]
  1> --------[test_0][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][0]
  1> --------[test_0][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][0], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][7]
  1> --------[test_0][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][7], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][3]
  1> --------[test_0][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][1]
  1> --------[test_0][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][5]
  1> --------[test_0][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_0][6]
  1> --------[test_0][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> ----shard_id [test_0][4]
  1> --------[test_0][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_8]
  1> ----shard_id [test_8][0]
  1> --------[test_8][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_8][1]
  1> --------[test_8][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_8][2]
  1> --------[test_8][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1>
  1> -- index [test_7]
  1> ----shard_id [test_7][4]
  1> --------[test_7][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_7][0]
  1> --------[test_7][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_7][3]
  1> --------[test_7][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_7][1]
  1> --------[test_7][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_7][2]
  1> --------[test_7][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1>
  1> -- index [test_6]
  1> ----shard_id [test_6][0]
  1> --------[test_6][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][3]
  1> --------[test_6][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][1]
  1> --------[test_6][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_6][2]
  1> --------[test_6][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_6][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1>
  1> -- index [test_5]
  1> ----shard_id [test_5][0]
  1> --------[test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> ----shard_id [test_5][3]
  1> --------[test_5][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_5][1]
  1> --------[test_5][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ----shard_id [test_5][2]
  1> --------[test_5][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1>
  1> routing_nodes:
  1> -----node_id[GQ6yYxmyRT-sfvT0cmuqQQ][V]
  1> --------[test_4][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_4][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_4][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_3][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_3][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_2][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_2][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_1][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][7], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][5], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_0][6], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_0][4], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_8][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_8][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_7][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_7][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_6][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][3], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][1], node[GQ6yYxmyRT-sfvT0cmuqQQ], [R], s[STARTED]
  1> --------[test_6][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_5][0], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> --------[test_5][2], node[GQ6yYxmyRT-sfvT0cmuqQQ], [P], s[STARTED]
  1> -----node_id[G4AEDzbrRae5BC_UD9zItA][V]
  1> --------[test_4][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][7], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_4][6], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_4][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_3][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_3][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_3][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_2][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_2][3], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][1], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][5], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_2][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_1][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][0], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][7], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][5], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_0][6], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_0][4], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_8][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][4], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_7][2], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][0], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_6][2], node[G4AEDzbrRae5BC_UD9zItA], [R], s[STARTED]
  1> --------[test_5][3], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> --------[test_5][1], node[G4AEDzbrRae5BC_UD9zItA], [P], s[STARTED]
  1> ---- unassigned
  1>
  1> tasks: (0):
  1>

[...]
```

--- a/core/src/main/java/org/elasticsearch/index/shard/IndexShard.java
+++ b/core/src/main/java/org/elasticsearch/index/shard/IndexShard.java
@@ -111,6 +111,7 @@ import java.io.IOException;
 import java.io.PrintStream;
 import java.nio.channels.ClosedByInterruptException;
 import java.util.Arrays;
+import java.util.EnumSet;
 import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.CopyOnWriteArrayList;
@@ -190,6 +191,8 @@ public class IndexShard extends AbstractIndexShardComponent {
     private final ShardPath path;
 
     private final IndexShardOperationCounter indexShardOperationCounter;
+
+    private EnumSet<IndexShardState> readAllowedStates = EnumSet.of(IndexShardState.STARTED, IndexShardState.RELOCATED, IndexShardState.POST_RECOVERY);
 
     @Inject
     public IndexShard(ShardId shardId, IndexSettingsService indexSettingsService, IndicesLifecycle indicesLifecycle, Store store, StoreRecoveryService storeRecoveryService,
@@ -953,8 +956,8 @@ public class IndexShard extends AbstractIndexShardComponent {
 
     public void readAllowed() throws IllegalIndexShardStateException {
         IndexShardState state = this.state; // one time volatile read
-        if (state != IndexShardState.STARTED && state != IndexShardState.RELOCATED) {
-            throw new IllegalIndexShardStateException(shardId, state, "operations only allowed when started/relocated");
+        if (readAllowedStates.contains(state) == false) {
+            throw new IllegalIndexShardStateException(shardId, state, "operations only allowed when shard state is one of " + readAllowedStates.toString());
         }
     }
 

--- a/core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java
+++ b/core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java
@@ -23,21 +23,29 @@ import com.google.common.base.Predicate;
 import org.apache.lucene.util.LuceneTestCase;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
-import org.elasticsearch.action.admin.cluster.state.ClusterStateResponse;
+import org.elasticsearch.action.admin.cluster.reroute.ClusterRerouteResponse;
+import org.elasticsearch.action.admin.indices.flush.FlushResponse;
+import org.elasticsearch.action.admin.indices.refresh.RefreshResponse;
+import org.elasticsearch.action.count.CountResponse;
 import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.index.IndexResponse;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.cluster.*;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
 import org.elasticsearch.cluster.block.ClusterBlock;
 import org.elasticsearch.cluster.block.ClusterBlockLevel;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.node.DiscoveryNodes;
 import org.elasticsearch.cluster.routing.DjbHashFunction;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.allocation.command.MoveAllocationCommand;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.logging.ESLogger;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.discovery.zen.ZenDiscovery;
@@ -48,6 +56,10 @@ import org.elasticsearch.discovery.zen.ping.ZenPing;
 import org.elasticsearch.discovery.zen.ping.ZenPingService;
 import org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing;
 import org.elasticsearch.discovery.zen.publish.PublishClusterStateAction;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.recovery.RecoverySource;
+import org.elasticsearch.indices.store.IndicesStoreIntegrationIT;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.test.ESIntegTestCase;
 import org.elasticsearch.test.InternalTestCluster;
@@ -55,7 +67,10 @@ import org.elasticsearch.test.discovery.ClusterDiscoveryConfiguration;
 import org.elasticsearch.test.disruption.*;
 import org.elasticsearch.test.junit.annotations.TestLogging;
 import org.elasticsearch.test.transport.MockTransportService;
-import org.elasticsearch.transport.*;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportService;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -812,7 +827,9 @@ public class DiscoveryWithServiceDisruptionsIT extends ESIntegTestCase {
     }
 
 
-    /** Test cluster join with issues in cluster state publishing * */
+    /**
+     * Test cluster join with issues in cluster state publishing *
+     */
     @Test
     public void testClusterJoinDespiteOfPublishingIssues() throws Exception {
         List<String> nodes = startCluster(2, 1);
@@ -919,6 +936,277 @@ public class DiscoveryWithServiceDisruptionsIT extends ESIntegTestCase {
         ensureStableCluster(3);
     }
 
+    /*
+     * Tests a visibility issue if a shard is in POST_RECOVERY
+     *
+     * When a user indexes a document, then refreshes and then a executes a search and all are successful and no timeouts etc then
+     * the document must be visible for the search.
+     *
+     * When a primary is relocating from node_1 to node_2, there can be a short time where both old and new primary
+     * are started and accept indexing and read requests. However, the new primary might not be visible to nodes
+     * that lag behind one cluster state. If such a node then sends a refresh to the index, this refresh request
+     * must reach the new primary on node_2 too. Otherwise a different node that searches on the new primary might not
+     * find the indexed document although a refresh was executed before.
+     *
+     * In detail:
+     * Cluster state 0:
+     * node_1: [index][0] STARTED   (ShardRoutingState)
+     * node_2: no shard
+     *
+     * 0. primary ([index][0]) relocates from node_1 to node_2
+     * Cluster state 1:
+     * node_1: [index][0] RELOCATING   (ShardRoutingState), (STARTED from IndexShardState perspective on node_1)
+     * node_2: [index][0] INITIALIZING (ShardRoutingState), (IndexShardState on node_2 is RECOVERING)
+     *
+     * 1. node_2 is done recovering, moves its shard to IndexShardState.POST_RECOVERY and sends a message to master that the shard is ShardRoutingState.STARTED
+     * Cluster state is still the same but the IndexShardState on node_2 has changed and it now accepts writes and reads:
+     * node_1: [index][0] RELOCATING   (ShardRoutingState), (STARTED from IndexShardState perspective on node_1)
+     * node_2: [index][0] INITIALIZING (ShardRoutingState), (IndexShardState on node_2 is POST_RECOVERY)
+     *
+     * 2. any node receives an index request which is then executed on node_1 and node_2
+     *
+     * 3. node_3 sends a refresh but it is a little behind with cluster state processing and still on cluster state 0.
+     * If refresh was a broadcast operation it send it to node_1 only because it does not know node_2 has a shard too
+     *
+     * 4. node_3 catches up with the cluster state and acks it to master which now can process the shard started message
+     *  from node_2 before and updates cluster state to:
+     * Cluster state 2:
+     * node_1: [index][0] no shard
+     * node_2: [index][0] STARTED (ShardRoutingState), (IndexShardState on node_2 is still POST_RECOVERY)
+     *
+     * master sends this to all nodes.
+     *
+     * 5. node_4 and node_3 process cluster state 2, but node_1 and node_2 have not yet
+     *
+     * If now node_4 searches for document that was indexed before, it will search at node_2 because it is on
+     * cluster state 2. It should be able to retrieve it with a search because the refresh from before was
+     * successful.
+     */
+    @Test
+    public void testReadOnPostRecoveryShards() throws Exception {
+        List<BlockClusterStateProcessing> clusterStateBlocks = new ArrayList<>();
+        try {
+            configureUnicastCluster(5, null, 1);
+            // we could probably write a test without a dedicated master node but it is easier if we use one
+            Future<String> masterNodeFuture = internalCluster().startMasterOnlyNodeAsync();
+            // node_1 will have the shard in the beginning
+            Future<String> node1Future = internalCluster().startDataOnlyNodeAsync();
+            final String masterNode = masterNodeFuture.get();
+            final String node_1 = node1Future.get();
+            logger.info("--> creating index [test] with one shard and zero replica");
+            assertAcked(prepareCreate("test").setSettings(
+                            Settings.builder().put(indexSettings())
+                                    .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                                    .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)
+                                    .put(IndexShard.INDEX_REFRESH_INTERVAL, -1))
+                            .addMapping("doc", jsonBuilder().startObject().startObject("doc")
+                                    .startObject("properties").startObject("text").field("type", "string").endObject().endObject()
+                                    .endObject().endObject())
+            );
+            ensureGreen("test");
+            logger.info("--> starting three more data nodes");
+            List<String> nodeNamesFuture = internalCluster().startDataOnlyNodesAsync(3).get();
+            final String node_2 = nodeNamesFuture.get(0);
+            final String node_3 = nodeNamesFuture.get(1);
+            final String node_4 = nodeNamesFuture.get(2);
+            logger.info("--> running cluster_health");
+            ClusterHealthResponse clusterHealth = client().admin().cluster().prepareHealth()
+                    .setWaitForNodes("5")
+                    .setWaitForRelocatingShards(0)
+                    .get();
+            assertThat(clusterHealth.isTimedOut(), equalTo(false));
+
+            logger.info("--> move shard from node_1 to node_2, and wait for relocation to finish");
+
+            // block cluster state updates on node_3 so that it only sees the shard on node_1
+            BlockClusterStateProcessing disruptionNode3 = new BlockClusterStateProcessing(node_3, getRandom());
+            clusterStateBlocks.add(disruptionNode3);
+            internalCluster().setDisruptionScheme(disruptionNode3);
+            disruptionNode3.startDisrupting();
+            // register a Tracer that notifies begin and end of a relocation
+            MockTransportService transportServiceNode2 = (MockTransportService) internalCluster().getInstance(TransportService.class, node_2);
+            CountDownLatch beginRelocationLatchNode2 = new CountDownLatch(1);
+            CountDownLatch endRelocationLatchNode2 = new CountDownLatch(1);
+            transportServiceNode2.addTracer(new StartRecoveryToShardStaredTracer(logger, beginRelocationLatchNode2, endRelocationLatchNode2));
+
+            // block cluster state updates on node_1 and node_2 so that we end up with two primaries
+            BlockClusterStateProcessing disruptionNode2 = new BlockClusterStateProcessing(node_2, getRandom());
+            clusterStateBlocks.add(disruptionNode2);
+            disruptionNode2.applyToCluster(internalCluster());
+            BlockClusterStateProcessing disruptionNode1 = new BlockClusterStateProcessing(node_1, getRandom());
+            clusterStateBlocks.add(disruptionNode1);
+            disruptionNode1.applyToCluster(internalCluster());
+
+            logger.info("--> move shard from node_1 to node_2");
+            // don't block on the relocation. cluster state updates are blocked on node_3 and the relocation would timeout
+            Future<ClusterRerouteResponse> rerouteFuture = internalCluster().client().admin().cluster().prepareReroute().add(new MoveAllocationCommand(new ShardId("test", 0), node_1, node_2)).setTimeout(new TimeValue(1000, TimeUnit.MILLISECONDS)).execute();
+
+            logger.info("--> wait for relocation to start");
+            // wait for relocation to start
+            beginRelocationLatchNode2.await();
+            // start to block cluster state updates on node_1 and node_2 so that we end up with two primaries
+            // one STARTED on node_1 and one in POST_RECOVERY on node_2
+            disruptionNode1.startDisrupting();
+            disruptionNode2.startDisrupting();
+            endRelocationLatchNode2.await();
+            final Client node3Client = internalCluster().client(node_3);
+            final Client node2Client = internalCluster().client(node_2);
+            final Client node1Client = internalCluster().client(node_1);
+            final Client node4Client = internalCluster().client(node_4);
+            logger.info("--> index doc");
+            logLocalClusterStates(node1Client, node2Client, node3Client,  node4Client);
+            assertTrue(node3Client.prepareIndex("test", "doc").setSource("{\"text\":\"a\"}").get().isCreated());
+            //sometimes refresh and sometimes flush
+            int refreshOrFlushType = randomIntBetween(1, 2);
+            switch (refreshOrFlushType) {
+                case 1: {
+                    logger.info("--> refresh from node_3");
+                    RefreshResponse refreshResponse = node3Client.admin().indices().prepareRefresh().get();
+                    assertThat(refreshResponse.getFailedShards(), equalTo(0));
+                    // the total shards is num replicas + 1 so that can be lower here because one shard
+                    // is relocating and counts twice as successful
+                    assertThat(refreshResponse.getTotalShards(), equalTo(2));
+                    assertThat(refreshResponse.getSuccessfulShards(), equalTo(2));
+                    break;
+                }
+                case 2: {
+                    logger.info("--> flush from node_3");
+                    FlushResponse flushResponse = node3Client.admin().indices().prepareFlush().get();
+                    assertThat(flushResponse.getFailedShards(), equalTo(0));
+                    // the total shards is num replicas + 1 so that can be lower here because one shard
+                    // is relocating and counts twice as successful
+                    assertThat(flushResponse.getTotalShards(), equalTo(2));
+                    assertThat(flushResponse.getSuccessfulShards(), equalTo(2));
+                    break;
+                }
+                default:
+                    fail("this is  test bug, number should be between 1 and 2");
+            }
+            // now stop disrupting so that node_3 can ack last cluster state to master and master can continue
+            // to publish the next cluster state
+            logger.info("--> stop disrupting node_3");
+            disruptionNode3.stopDisrupting();
+            rerouteFuture.get();
+            logger.info("--> wait for node_4 to get new cluster state");
+            // wait until node_4 actually has the new cluster state in which node_1 has no shard
+            assertBusy(new Runnable() {
+                @Override
+                public void run() {
+                    ClusterState clusterState = node4Client.admin().cluster().prepareState().setLocal(true).get().getState();
+                    // get the node id from the name. TODO: Is there a better way to do this?
+                    String nodeId = null;
+                    for (RoutingNode node : clusterState.getRoutingNodes()) {
+                        if (node.node().name().equals(node_1)) {
+                            nodeId = node.nodeId();
+                        }
+                    }
+                    assertNotNull(nodeId);
+                    // check that node_1 does not have the shard in local cluster state
+                    assertFalse(clusterState.getRoutingNodes().routingNodeIter(nodeId).hasNext());
+                }
+            });
+
+            logger.info("--> run count from node_4");
+            logLocalClusterStates(node1Client, node2Client, node3Client, node4Client);
+            CountResponse countResponse = node4Client.prepareCount("test").setPreference("local").get();
+            assertThat(countResponse.getCount(), equalTo(1l));
+            logger.info("--> stop disrupting node_1 and node_2");
+            disruptionNode2.stopDisrupting();
+            disruptionNode1.stopDisrupting();
+            // wait for relocation to finish
+            logger.info("--> wait for relocation to finish");
+            clusterHealth = client().admin().cluster().prepareHealth()
+                    .setWaitForRelocatingShards(0)
+                    .get();
+            assertThat(clusterHealth.isTimedOut(), equalTo(false));
+        } catch (AssertionError e) {
+            for (BlockClusterStateProcessing blockClusterStateProcessing : clusterStateBlocks) {
+                blockClusterStateProcessing.stopDisrupting();
+            }
+            throw e;
+        }
+    }
+
+    /**
+     * This Tracer can be used to signal start of a recovery and shard started event after translog was copied
+     */
+    public static class StartRecoveryToShardStaredTracer extends MockTransportService.Tracer {
+        private final ESLogger logger;
+        private final CountDownLatch beginRelocationLatch;
+        private final CountDownLatch sentShardStartedLatch;
+
+        public StartRecoveryToShardStaredTracer(ESLogger logger, CountDownLatch beginRelocationLatch, CountDownLatch sentShardStartedLatch) {
+            this.logger = logger;
+            this.beginRelocationLatch = beginRelocationLatch;
+            this.sentShardStartedLatch = sentShardStartedLatch;
+        }
+
+        @Override
+        public void requestSent(DiscoveryNode node, long requestId, String action, TransportRequestOptions options) {
+            if (action.equals(RecoverySource.Actions.START_RECOVERY)) {
+                logger.info("sent: {}, relocation starts", action);
+                beginRelocationLatch.countDown();
+            }
+            if (action.equals(ShardStateAction.SHARD_STARTED_ACTION_NAME)) {
+                logger.info("sent: {}, shard started", action);
+                sentShardStartedLatch.countDown();
+            }
+        }
+    }
+
+    private void logLocalClusterStates(Client... clients) {
+        int counter = 1;
+        for (Client client : clients) {
+            ClusterState clusterState = client.admin().cluster().prepareState().setLocal(true).get().getState();
+            logger.info("--> cluster state on node_{} {}", counter, clusterState.prettyPrint());
+            counter++;
+        }
+    }
+
+    /**
+     * This test creates a scenario where a primary shard (0 replicas) relocates and is in POST_RECOVERY on the target
+     * node but already deleted on the source node. Search request should still work.
+     */
+    @Test
+    public void searchWithRelocationAndSlowClusterStateProcessing() throws Exception {
+        configureUnicastCluster(3, null, 1);
+        Future<String> masterNodeFuture = internalCluster().startMasterOnlyNodeAsync();
+        Future<String> node_1Future = internalCluster().startDataOnlyNodeAsync();
+
+        final String node_1 = node_1Future.get();
+        final String masterNode = masterNodeFuture.get();
+        logger.info("--> creating index [test] with one shard and on replica");
+        assertAcked(prepareCreate("test").setSettings(
+                        Settings.builder().put(indexSettings())
+                                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+                                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0))
+        );
+        ensureGreen("test");
+
+        Future<String> node_2Future = internalCluster().startDataOnlyNodeAsync();
+        final String node_2 = node_2Future.get();
+        List<IndexRequestBuilder> indexRequestBuilderList = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            indexRequestBuilderList.add(client().prepareIndex().setIndex("test").setType("doc").setSource("{\"int_field\":1}"));
+        }
+        indexRandom(true, indexRequestBuilderList);
+        SingleNodeDisruption disruption = new BlockClusterStateProcessing(node_2, getRandom());
+
+        internalCluster().setDisruptionScheme(disruption);
+        MockTransportService transportServiceNode2 = (MockTransportService) internalCluster().getInstance(TransportService.class, node_2);
+        CountDownLatch beginRelocationLatch = new CountDownLatch(1);
+        CountDownLatch endRelocationLatch = new CountDownLatch(1);
+        transportServiceNode2.addTracer(new IndicesStoreIntegrationIT.ReclocationStartEndTracer(logger, beginRelocationLatch, endRelocationLatch));
+        internalCluster().client().admin().cluster().prepareReroute().add(new MoveAllocationCommand(new ShardId("test", 0), node_1, node_2)).get();
+        // wait for relocation to start
+        beginRelocationLatch.await();
+        disruption.startDisrupting();
+        // wait for relocation to finish
+        endRelocationLatch.await();
+        // now search for the documents and see if we get a reply
+        assertThat(client().prepareCount().get().getCount(), equalTo(100l));
+    }
+
     @Test
     public void testIndexImportedFromDataOnlyNodesIfMasterLostDataFolder() throws Exception {
         // test for https://github.com/elastic/elasticsearch/issues/8823
@@ -932,6 +1220,7 @@ public class DiscoveryWithServiceDisruptionsIT extends ESIntegTestCase {
         ensureGreen();
 
         internalCluster().restartNode(masterNode, new InternalTestCluster.RestartCallback() {
+            @Override
             public boolean clearData(String nodeName) {
                 return true;
             }

